### ISSUE TITLE
Save before start is broken

### ISSUE THREAD

**Issue by MarByteBeep** [microsoft/vscode#125259]
Issue Type: <b>Bug</b>

I have "Save before start" enabled in the config, yet it doesn't seem to notice the saved changes until I run the app again. Steps to repro:

* change launch.json 
* press F5
* observe that the changes are not in effect
* press F5 again
* changes are now in effect

VS Code version: Code 1.56.2 (054a9295330880ed74ceaedda236253b4f39a335, 2021-05-12T17:13:13.157Z)
OS version: Windows_NT x64 10.0.19042

**Comment by isidorn** [comment]
I agree that we do not automatically save `launch.json` before debug starts. The reason for this is that we simply first need to have a launch configuration before debugging starts and thus we save.
I will leave this open for now and we might fix it in the future.
Strangely you are the first user to find this limitation.

**Comment by MarByteBeep** [comment]
Thanks! Well the weird thing is that it actually was saved upon hitting F5, however the debug session still uses the earlier version. It's indeed very strange that no-one noticed this before, unless of course it used to work fine and got broken recently. Thanks for your reply!

**Comment by isidorn** [comment]
Yeah, I think it was never working... I do agree it is a bit strange and can be confusing. We should fix.

**Comment by asafmanor** [comment]
Hi. Having this problem as well. I believe it started a few months ago - before that, there was no need to double press the debug button in order for it to update.

**Comment by weinand** [comment]
@isidorn is it really necessary to "save" the launch.json? VS Code should be able to pick up the latest version of it without having it to read from disk, correct?

**Comment by MarByteBeep** [comment]
@asafmanor exactly, I'm pretty sure I've been using this for months and I would've noticed this problem earlier if it never worked properly. I think this got broken recently.

**Comment by weinand** [comment]
@isidorn has something changed in this area for the "restart" fix?

**Comment by isidorn** [comment]
No, nothing changed here from our side.
We are reading the configuration here https://github.com/microsoft/vscode/blob/main/src/vs/workbench/contrib/debug/browser/debugCommands.ts#L394

So the solution would be to trigger another save before we read the configuration. Unfortunetly this does not work nicely without a timeout since it takes some time for the `configurationService` to pick up the change.
Or that we lazily read the configuration. To lazily read the configuration is a bigger refactoring.

Users that know this was working before can you please let me know in what vscode version was this working?

**Comment by MarByteBeep** [comment]
@isidorn I just narrowed it down for you: 1.54.3 seems to work fine, 1.55.2 does not. So it got borked between February and March of this year.

**Comment by isidorn** [comment]
Will not have time for this milestone, next milestone I am on vacation -> assigning to August.
If somebody provides a PR that fixes this I am open for a review. 
Code pointer https://github.com/microsoft/vscode/blob/main/src/vs/workbench/contrib/debug/browser/debugCommands.ts#L406

**Comment by MarByteBeep** [comment]
Awesome! Thanks and confirmed!
